Adding kind to external link

artemruts opened the pull request:
Fixes #14979

### Description

`ExternalLink` will expose `serviceKind` nullable enum instead of `serviceType` string.

artemruts commented:
@felixfbecker Can you please take a look and let me know if this is in the ballpark of what you wanted?
I'll still need to fix a few tests and update snapshots.

felixfbecker commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -4164,10 +4164,10 @@ type ExternalLink {
     url: String!
     """
-    The type of external service, such as "github", or null if unknown/unrecognized. This is used solely for
+    The kind of external service, such as "GITHUB", or "OTHER" if unknown/unrecognized. This is used solely for
```
If it's `OTHER` when it's unrecognized, in which cases is it `null`? (should be documented, or if it's never `null`, it should be non-nullable)

artemruts replied:
Felix, I changed that comment and re-evaluated how it'll work. Kind will be of ExternalServicesKind enum which includes `OTHER` (as I understand it's a known generic external service) + `null` in case it's some other value.

felixfbecker reviewed (commented):
LGTM besides my comment!

artemruts commented:
> Nice! Worth a changelog entry I think, since we changed "github" to "GITHUB", in case some clients depend on its value.

@unknwon Is there a way to check if those APIs  where `type` became `kind` are not used by clients?

eseliger commented on cmd/frontend/graphqlbackend/externallink/resolver.go:
```diff
@@ -9,19 +9,22 @@ import "fmt"
+func (r *Resolver) ServiceKind() *string {
+	if r.serviceKind == "" {
 		return nil
+	} else {
+		return &r.serviceKind
 	}
```
No else branch required :) 

```suggestion
	}
	return &r.serviceKind
```

eseliger commented on client/web/src/repo/actions/InstallBrowserExtensionPopover.tsx:
```diff
@@ -61,9 +62,9 @@ export const InstallBrowserExtensionPopover: React.FunctionComponent<Props> = ({
                         <p className="py-3">
                             Install Sourcegraph browser extension to add code intelligence{' '}
-                            {serviceType === 'phabricator'
+                            {serviceKind === ExternalServiceKind.PHABRICATOR
```
❤️

eseliger commented on client/web/src/repo/actions/InstallBrowserExtensionAlert.tsx:
```diff
@@ -15,21 +17,26 @@ interface Props {
 /** Code hosts the browser extension supports */
-const supportedServiceTypes = new Set<string>(['github', 'gitlab', 'phabricator', 'bitbucketServer'])
+const supportedServiceTypes = new Set<string>([
+    GQL.ExternalServiceKind.GITHUB,
+    GQL.ExternalServiceKind.GITLAB,
+    GQL.ExternalServiceKind.PHABRICATOR,
+    GQL.ExternalServiceKind.BITBUCKETSERVER,
```
These types are deprecated and they should ideally be imported from '../..graphql-operations' directly (without `GQL.` namespacing)

eseliger commented on client/web/src/repo/actions/InstallBrowserExtensionAlert.tsx:
```diff
@@ -1,7 +1,9 @@
 import CloseIcon from 'mdi-react/CloseIcon'
 import ExportIcon from 'mdi-react/ExportIcon'
-import { serviceTypeDisplayNameAndIcon } from './GoToCodeHostAction'
+
```
```suggestion
```

artemruts replied:
Sorry, old habits, I like to separate 3rd party/global package imports and local ones :) I'll stop that!

eseliger commented on client/web/src/repo/actions/GoToCodeHostAction.tsx:
```diff
@@ -16,6 +16,7 @@ import GitlabIcon from 'mdi-react/GitlabIcon'
 import { eventLogger } from '../../tracking/eventLogger'
 import { InstallBrowserExtensionPopover } from './InstallBrowserExtensionPopover'
 import { useLocalStorage } from '../../util/useLocalStorage'
+import * as GQL from '../../../../shared/src/graphql/schema'
```
can you please use the types from `graphql-operations`, which are the newer ones and these should go away at some point, thanks :)

eseliger commented on client/web/src/integration/graphQlResponseHelpers.ts:
```diff
@@ -7,6 +7,8 @@ import {
     ResolveRevResult,
 } from '../graphql-operations'
 
+import { ExternalServiceKind } from '../../../shared/src/graphql/schema'
```
```suggestion
ExternalServiceKind,
} from '../graphql-operations'
```

eseliger reviewed (approved):
Thank you this is a great change! I do share the concern about breaking the API without further notice, though. At least an entry to the changelog should be done, and if we feel extra fancy we might consider keeping `type` around until the 20th, and just remove the resolver after that date. Not sure.

artemruts commented:
> Thank you this is a great change! I do share the concern about breaking the API without further notice, though. At least an entry to the changelog should be done, and if we feel extra fancy we might consider keeping `type` around until the 20th, and just remove the resolver after that date. Not sure.

@eseliger I think I'll add the `type` back and add a deprecation notice to play it safe 👍

artemruts commented:
@unknwon Thanks for clarifying about the client scripts! I'm still not sure on what level they operate Web vs API calls. 

I did the following:
- added `serviceType` back to the resolver
- marked `serviceType` as deprecated
- added notes in changes log

![deprecated](https://user-images.githubusercontent.com/1319181/103826847-a79d4700-5045-11eb-9ca8-15a8e694206d.png)

eseliger commented:
Awesome @artemruts, thanks!
